pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -506,7 +515,33 @@ func (configgen *ConfigGeneratorImpl) buildSidecarOutboundListeners(env model.En
 		}
 	}
 
-	return append(tcpListeners, httpListeners...)
+	listeners := append(tcpListeners, httpListeners...)
+
+	// trim conflicting filterchains
+	// FIXME: ugly hack
+	for _, l := range listeners {
+		filterChainMatches := make(map[string]bool)
+
+		trimmedFilterChains := make([]listener.FilterChain, 0, len(l.FilterChains))
+		for _, filterChain := range l.FilterChains {
+			var key string
+			if filterChain.FilterChainMatch != nil {
+				sniDomains := make([]string, len(filterChain.FilterChainMatch.SniDomains))
+				copy(sniDomains, filterChain.FilterChainMatch.SniDomains)
+				sort.Strings(sniDomains)
+				key = strings.Join(sniDomains, ",") // sni domains is the only thing set in filterchainmatch rn
+			}
+			if !filterChainMatches[key] {
```

Comment:
this should go into the if block above.

Reply:
Line 416?
`if currentListener, exists = listenerMap[listenerMapKey]; exists {`